Sources/StreamChat/StateLayer/Pagination/PaginatedChannelsLoader.swift

```diff
@@ -0,0 +1,43 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+@available(iOS 13.0, *)
+struct PaginatedChannelsLoader {
+    let query: ChannelListQuery
+    let chatClientConfig: ChatClientConfig
+    let channelListUpdater: ChannelListUpdater
+    let channelRepository: ChannelRepository
+    
+    func loadChannels(to state: ChatListState, pagination: Pagination) async throws -> [ChatChannel] {
+        let query = self.query.withPagination(pagination)
+        let payloadChannels = try await channelListUpdater.update(channelListQuery: query)
+        // Optimization: remove the database fetch and instead replicate filtering and sorting defined by ChannelDTO.channelListFetchRequest(query:chatClientConfig:). Currently we can't ensure that payloadChannels is sorted the same way.
+        let newSortedChannels = try await channelRepository.channels(
+            for: payloadChannels.map(\.cid),
+            query: query,
+            config: chatClientConfig
+        )
+        let result = await state.orderedChannels.withInsertingPaginated(newSortedChannels)
+        await state.setSortedChannels(result, canLoadMore: payloadChannels.count < pagination.pageSize)
+        
+        return payloadChannels
```

Comment:
If we move away from FRC driven data source then here we can optimise pagination to only use API requests and its payloads and skipping some of the DB writes and reads. This is for future iterations.